### PR TITLE
Magic enum format in a separated header file 

### PR DESCRIPTION
**Pull request by schaumb**
Based on https://github.com/Neargye/magic_enum/issues/151

### CODE REVIEW

**Comment by schaumb** [comment]
It needs some documentation (?) and test(s) ✓.

**Comment by Neargye** [comment]
It looks really good with separate headers.

**Comment by schaumb** [comment]
!!! This formatter uses std::string_view . If someone defines a different string_view, this will not work.

To resolve this problem one of the following is necessary:
- apply some conversion to std::string_view ✓ (`.data()`, `.size()` )
- ~~require its own formatter for the user-defined string_view~~ ✗
- ~~prohibit to use of this header~~ ✗

**Comment by Neargye** [comment]
Is it possible to use a  char* formatter?

**Comment by schaumb** [comment]
!!! Another problem is that this solution does not handle the enum flags.
- ~~one solution would be a unique format character, which enables to format of enum flags output, like `{...f}` ( 'f' character on the format string end can be a mean that handles this enum as a flag).~~ ✗
- or this can be the default behavior if the enum is a flag ✓

**Comment by schaumb** [comment]
> Is it possible to use a char* formatter?

string_view data is not required to end with `'\0'`, so the `const char*` formatter cannot be used

File: include/magic_enum_format.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  constexpr bool enum_format_enabled() noexcept {
+    return MAGIC_ENUM_DEFAULT_ENABLE_ENUM_FORMAT;
+  }
+} // magic_enum::customize
+
+#include <format>
+
+template<typename E>
+struct std::formatter<E, std::enable_if_t<std::is_enum_v<E> && magic_enum::customize::enum_format_enabled<E>(), char>> : std::formatter<std::string_view, char> {
+  auto format(E e, format_context& ctx) {
```
Comment (Neargye) [inline include/magic_enum_format.hpp:58]: static_assert(std::is_constructible<std::string_view, decltype(magic_enum::enum_name(e)));
to fix https://github.com/Neargye/magic_enum/pull/185#issuecomment-1126173671

File: include/magic_enum_format.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    if (auto name = magic_enum::enum_name(e); name.empty()) {
+      constexpr auto type_name = magic_enum::enum_type_name<E>();
+      throw std::format_error("Type of " + std::string{type_name.data(), type_name.size()} + " enum value: " +
+            std::to_string(magic_enum::enum_integer(e)) + " is not exists.");
+    } else {
+      return this->std::formatter<std::string_view, char>::format(name, ctx);
```
Comment (Neargye) [inline include/magic_enum_format.hpp]: std::string_view{name.data(), name.size()}

File: include/magic_enum_format.hpp (hunk 1 above)
Comment (schaumb) [reply include/magic_enum_format.hpp:58]: this check is not needed, because the other, `std::string_view{name.data(), name.size()}` conversion is enough

**Comment by Timple** [comment]
Awesome job!

The [libfmt](https://fmt.dev/latest/api.html#format-api) library provides this functionality for `c++17`. This example works.

Would it make sense to combine this with this file?

```c++
template<typename E>
struct fmt::formatter<E, std::enable_if_t<std::is_enum_v<E>, char>> : fmt::formatter<std::string_view, char> {
  auto format(E e, format_context& ctx) {
    if constexpr (magic_enum::detail::is_flags_v<E>) {
      if (auto name = magic_enum::enum_flags_name(e); !name.empty()) {
        return this->fmt::formatter<std::string_view, char>::format(std::string_view{name.data(), name.size()}, ctx);
      }
    } else {
      if (auto name = magic_enum::enum_name(e); !name.empty()) {
        return this->fmt::formatter<std::string_view, char>::format(std::string_view{name.data(), name.size()}, ctx);
      }
    }
    constexpr auto type_name = magic_enum::enum_type_name<E>();
    throw std::domain_error("Type of " + std::string{type_name.data(), type_name.size()} + " enum value: " +
                            std::to_string(magic_enum::enum_integer(e)) + " is not exists.");
  }
};
```

**Comment by Neargye** [comment]
@Timple please create new issues

**Pull request merged**